### DISCUSSION TITLE
Reached limit of HTTP authorization service throughput

### DISCUSSION THREAD

**Discussion by ddellarocca** [emqx/emqx#12618] [Q&A]
Hi,
I'm developing an HTTP authorization service that will have to validate the requests coming from EMQX.
EMQX is running in a 3 node cluster in K8S with each replica having 7 CPU and 29GiB of memory.
Moreover I have to validate spikes of more than 30k publish requests per second without leveraging the EMQX authorization cache since the outcome can change.
Each replica of my HTTP authorization service can handle 10 concurrent requests and overall 2k req/sec, so I've deployed 20 replicas of it and I've configured the HTTP Pool Size to 200. Testing the throughput of the deployment of my authorization server with apache benchmark i get around 38k req/sec.
The problem is that EMQX is not able to reach those throughput and is stuck at 9.7k tps so it ingests only 9.7k messages/sec which is not enough for my use case, with the authorization service disabled i can reach more than 30k messages/sec of ingestion.
What else can I change or try?

**Comment by ddellarocca** [comment]
Forgot to point out that the version of EMQX is 5.0.11

**Reply by kjellwinblad** [reply]
Thank you for an interesting question. Maybe you could try to change the Connection Pool Size and the HTTP Pipelining settings in a systematic way to see how these settings affect the throughput? Please let us know how it goes.

**Reply by ddellarocca** [reply]
Hi, I've done a lot more testing in the mean time, trying to tweak the connection pool size and http pipelining doesn't change the result.
On the other hand I found that the bottleneck Is emqx itself, during my tests it consumed all the CPU available, pinning it to 100%. Adding another instance will increase the throughput by ~3,2k TPS, so with 7 CPU cores each instance can validate around 3,2k messages per second.
I don't know if it is related to the http client itself or the internal process that validates the request, to test this out I'll set up the validation with redis in order to use a binary TCP connection.
I'll let you know the results

**Reply by ddellarocca** [reply]
I've done other tests and the results are:

- 10k TPS per instance with ACL file authorization
- 7K TPS per instance with REDIS authorization
- 3k TPS per instance with HTTP authorization

The tests are done with Jmeter running in a docker container in the same k8s cluster as EMQX. The Jmeter test is structured to create 10k threads that will:

- connect to the broker with an clientid that is a static string with the thread number to make it unique
- publish a message and wait 1ms looped over 1000 times
- disconnect from the broker

On the other hand the authorization services are configured this way:

- ACL: is the default one 
- REDIS: it has an item for each clientid containing a single key as the topic in which the Jmeter client will publish and its value set as `publish`. Its throughput is around 60k HGETALL/s
- HTTP: it always return allow, configured for 60k req/s, pool size set to 200 and http pipelining set to 100 (tried to tweak those values without success, but number lower than those will decrease performance)

During all those tests the bottleneck were EMQX since it was using all the CPU assigned to it (7 CPU per instance) while REDIS (around 30% CPU usage) and HTTP server (around 20% CPU usage) were underutilized.

**Reply by kjellwinblad** [reply]
Thanks for the additional information. Not sure if I understand the test set up. Is it correct that each client thread does the following:

LOOP:
1. Connect as an MQTT client to EMQX
2. Publish 1000 messages with 1 ms between each message
3. Disconnect from EMQX

In that case do you count each loop body execution as a transaction or do you count each message publish as a transaction? Does all messages generate an authorization request?

**Reply by ddellarocca** [reply]
The test setup is different actually, I'll paste an image to make it more clear:
![image](https://github.com/emqx/emqx/assets/29358615/fac81b25-a083-4143-9fcc-a7a60f986bba)


So:

- 10000 threads are spawned in parallel, each thread will
    - connect to the broker
    - loop 1000 times
        - publish 1 message
        - wait 1ms
    - disconnect

> In that case do you count each loop body execution as a transaction or do you count each message publish as a transaction? 

Each message published is a transaction because each publish will trigger an authorization request to the authorizer

> Does all messages generate an authorization request?

Yes, since I can't use EMQX authorization cache in my specific use case

**Reply by kjellwinblad** [reply]
Ok, thanks. Now it is very clear. It indeed seems like EMQX's HTTP authorization module is a little bit slow.

Is caching not acceptable at all in your scenario or could a cache that is cleared every few seconds be an option for you?

Could an option be to scale the cluster to more than three nodes to reach the acceptable throughput be an option?

To figure out if there is some opportunity for optimization, I think we would need to do some profiling and look at the code for optimizations. First, I will discuss with the EMQX team to see if they have some idea about how to speed it up and if these number look unreasonable to them.

**Reply by ddellarocca** [reply]
The problem in my use case is caching the deny results (a request can be temporary denied but shouldn't be all the time, it depends on the status of other components in my architecture).
For the time being I'll just scale up the cluster to match my desired throughput, but still is not optimal.

Thanks for the feeback.

**Comment by zmstone** [comment]
Hi @ddellarocca 
Thank you for sharing your tests with us.

I do not have a direct answer to why HTTP authz is slow in your tests.

Sharing some thoughts here:

* Maybe consider upgrading to 5.5.1 (latest release when writing this reply), there had been quite a lot changes since 5.0
~~* The most performant authorization solution is probably JWT, `acl.conf` file comes the the next if there isn't a lot of rules in the file.
   If ACL file being the only authorization source gives you 10K TPS, then I am actually a bit surprised that you can get up to 9.7K TPS using HTTP authorization. This also hints maybe the limiting factor is not the HTTP server or the pool size or the specific configurations etc.~~
* We are about to support HTTP authentication response with ACL rules (similar to what JWT's ACL claim), which will essentially deprecate most of the real-time authorization requests towards a HTTP server.

JWT ACL rules for refernece: https://www.emqx.io/docs/en/latest/access-control/authn/jwt.html#new-format

**Reply by zmstone** [reply]
Sorry, I had to re-read your message to realize that ACL.conf's 10K is per node, and HHTP's 9.7K is cluster-wide.
edited my reply above.

**Reply by ddellarocca** [reply]
Hi @zmstone,

> Maybe consider upgrading to 5.5.1 (latest release when writing this reply), there had been quite a lot changes since 5.0

Yes I'm considering it but it is not something that I can do in the next coming months.

> The most performant authorization solution is probably JWT, acl.conf file comes the the next if there isn't a lot of rules in the file.
    If ACL file being the only authorization source gives you 10K TPS, then I am actually a bit surprised that you can get up to 9.7K TPS using HTTP authorization. This also hints maybe the limiting factor is not the HTTP server or the pool size or the specific configurations etc.

With ACL I'm getting 10k TPS each instance so with 3 instances I get 30k TPS, with HTTP I get 9,7k TPS for 3 instances combined so 3,2k TPS per instance.

> We are about to support HTTP authentication response with ACL rules (similar to what JWT's ACL claim), which will essentially deprecate most of the real-time authorization requests towards a HTTP server.

The problem is that I can't change the clients, I can only authorize their actions.

**Reply by zmstone** [reply]
> The problem is that I can't change the clients, I can only authorize their actions.

If your clients are using HTTP authentication, then the upgrade do not require changes in clients when we have the ACL rules supported in authentication response.

**Reply by ddellarocca** [reply]
My clients are using the built-in authentication and they are connected using MQTT V5 protocol with username and password. The authorization is based on the clientid.

**Comment by zmstone** [comment]
Hi again

If the systems do not seem overloaded, the limiting factor is maybe latency.
e.g. if it takes around 1s to respond each PULBISH, then the overall TPS would be around 10K (since there are 10K concurrent clients),
increased latency can result in less overall throughput.

if you only want to test throughput, but not so concerned about latency, maybe you can try 30k concurrent clients instead.
(also keep an eye on the jmeter runner hosts' cpu utilization etc, maybe 30k clients is too much and slows down the clients).

Or maybe try to increase inflight window size.

**Reply by ddellarocca** [reply]
Hi again,

The limiting factor in my test were the CPU utilization on the node assigned to EMQX (on the EMQX node there weren't any other workloads), in fact in every tests EMQX was pinning the CPU utilization of the node to 100%. Adding an extra node increased the throughput of the whole cluster (https://github.com/emqx/emqx/discussions/12618#discussioncomment-8678396). My concern basically is the 70% drop in performance between ACL and HTTP.

The latency was negligible since the tests were done in the same cluster as EMQX.

**Reply by zmstone** [reply]
sorry for missing this part as well.

is it https ?

**Reply by zmstone** [reply]
also, does server keep connection alive ?

**Reply by ddellarocca** [reply]
It is HTTP and server keep the connection alive

**Comment by ddellarocca** [comment]
Hi, small update on this, tested with EMQX 5.5.1 and it seems a little more efficient, I can get:
- 10k TPS per instance with ACL file authorization
- 7.5K TPS per instance with REDIS authorization
- 3.5k TPS per instance with HTTP authorization
